src/analysis/normed/ring/seminorm_constructions.lean

```diff
@@ -0,0 +1,337 @@
+∀ (a : α) {n : ℕ} (hn : 1 ≤ n), f (a ^ n) = (f a) ^ n
+
+/-- A function `f : α → β` is nonarchimedean if it satisfies the inequality
+  `f (a + b) ≤ max (f a) (f b)` for all `a, b ∈ α`. -/
+def is_nonarchimedean {α : Type*} [has_add α] {β : Type*} [linear_order β] (f : α → β) : Prop :=
+∀ r s, f (r + s) ≤ max (f r) (f s)
+
+end defs
+
+section seminorm_from_const
+/- In this section we prove Proposition 1.3.2/2 in
+[S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]. -/
+
+section ring
+
+variables {R : Type*} [comm_ring R] (c : R) (f : ring_seminorm R) (hf1 : f 1 ≤ 1)
+  (hc : 0 ≠ f c) (hpm : is_pow_mul f)
+
+/-- For a ring seminorm `f` on `R` and `c ∈ R`, the sequence given by `(f (x * c^n))/((f c)^n)`. -/
+def seminorm_from_const_seq (x : R) : ℕ → ℝ :=
+λ n, (f (x * c^n))/((f c)^n)
+
+lemma seminorm_from_const_nonneg (x : R) (n : ℕ) : 0 ≤ seminorm_from_const_seq c f x n :=
+div_nonneg (map_nonneg f (x * c ^ n)) (pow_nonneg (map_nonneg f c) n)
+
+lemma seminorm_from_const_is_bounded (x : R) :
+  bdd_below (set.range (seminorm_from_const_seq c f x)) :=
+begin
+  use 0,
+  rw mem_lower_bounds,
+  rintros r ⟨n, hn⟩,
+  rw ← hn,
+  exact seminorm_from_const_nonneg c f x n,
+end
+
+variable {f}
+
+lemma seminorm_from_const_seq_zero (hf : f 0 = 0) :
+  seminorm_from_const_seq c f 0 = 0 :=
+begin
+  simp only [seminorm_from_const_seq],
+  ext n,
+  rw [zero_mul, hf, zero_div],
+  refl,
+end
+
+variable {c}
+
+include hc hpm
+
+lemma seminorm_from_const_seq_one (n : ℕ) (hn : 1 ≤ n) :
+  seminorm_from_const_seq c f 1 n = 1 :=
+begin
+  simp only [seminorm_from_const_seq],
+  rw [one_mul, hpm _ hn, div_self (pow_ne_zero n (ne.symm hc))],
+end
+
+include hf1
+
+lemma seminorm_from_const_seq_antitone (x : R) :
+  antitone (seminorm_from_const_seq c f x) :=
+begin
+  intros m n hmn,
+  simp only [seminorm_from_const_seq],
+  nth_rewrite 0 ← nat.add_sub_of_le hmn,
+  rw [pow_add, ← mul_assoc],
+  have hc_pos : 0 < f c := lt_of_le_of_ne (map_nonneg f _) hc,
+  apply le_trans ((div_le_div_right (pow_pos hc_pos _)).mpr (map_mul_le_mul f _ _)),
+  by_cases heq : m = n,
+  { have : n - m = 0,
+    { rw heq, exact nat.sub_self n, },
+    rw [this, heq, div_le_div_right (pow_pos hc_pos _), pow_zero],
+    conv_rhs{rw ← mul_one (f (x * c ^ n))},
+    exact mul_le_mul_of_nonneg_left hf1 (map_nonneg f _) },
+  { have h1 : 1 ≤ n - m,
+    { rw [nat.one_le_iff_ne_zero, ne.def, nat.sub_eq_zero_iff_le, not_le],
+    exact lt_of_le_of_ne hmn heq,},
+    rw [hpm c h1, mul_div_assoc, div_eq_mul_inv, pow_sub₀ _ (ne.symm hc) hmn, mul_assoc,
+      mul_comm (f c ^ m)⁻¹, ← mul_assoc (f c ^ n), mul_inv_cancel (pow_ne_zero n (ne.symm hc)),
+      one_mul, div_eq_mul_inv] }
+end
+
+/-- The limit of the sequence `seminorm_from_const`. -/
+def seminorm_from_const_seq_lim (x : R) : ℝ :=
+classical.some (real.exists_is_glb (set.range_nonempty (seminorm_from_const_seq c f x))
+  (seminorm_from_const_is_bounded c f x))
```

Comment:
Can this be stated with `⨆ n, seminorm_from_const_seq c f x`?

Reply:
It can be stated as `⨅ n, seminorm_from_const_seq c f x n`; I have changed it to that. Thanks!